File: SPOS/Python/7.LRU.py
class Page:
    def __init__(self, page_number):
        self.page_number = page_number
        self.referenced = 0

def lru(pages, page_frames):
    page_table = {}
    page_order = []

    page_faults = 0
    page_hits = 0

    for page_number in pages:
        if page_number in page_table:
            page_hits += 1
            page_table[page_number].referenced += 1
            # Move the accessed page to the end to represent it's the most recently used
            page_order.remove(page_table[page_number])
            page_order.append(page_table[page_number])
        else:
            page_faults += 1

            if len(page_table) == page_frames:
                # Remove the least recently used page (at the front of the list)
                removed_page = page_order.pop(0)
                del page_table[removed_page.page_number]

            new_page = Page(page_number)
            page_table[page_number] = new_page
            page_order.append(new_page)

        print(f"Cache after inserting page {page_number}: {[page.page_number for page in page_order]}")
        print(f"Hits: {page_hits}, Faults: {page_faults}\n")

    total_hits_and_faults = {"Hits": page_hits, "Faults": page_faults}
    return total_hits_and_faults

if __name__ == "__main__":
    pages_input = input("Enter the sequence of pages (space-separated): ")
    pages = [int(page) for page in pages_input.split()]

    page_frames = int(input("Enter the number of page frames: "))

    total_hits_and_faults = lru(pages, page_frames)

    print(f"Total Hits: {total_hits_and_faults['Hits']}, Total Faults: {total_hits_and_faults['Faults']}")
    print(f"Hit Ratio: {total_hits_and_faults['Hits'] / len(pages)}")

File: SPOS/CPP/7.LRU.cpp
#include <iostream>
#include <list>
#include <unordered_set>
#include <vector>

using namespace std;

void simulateLRUPageReplacement(const vector<int>& pageRequests, int frameCount) {
    list<int> pageFrames;
    unordered_set<int> pageSet;
    int pageFaults = 0;
    int pageHits = 0;

    cout << "+-----+--------------+----------------+----------------+" << endl;
    cout << "| Seq | Page Request | Page Frames     | Page Faults    |" << endl;
    cout << "+-----+--------------+----------------+----------------+" << endl;

    for (int i = 0; i < pageRequests.size(); i++) {
        int page = pageRequests[i];
        cout << "|  " << i + 1 << "  |      " << page << "       |";

        if (pageSet.find(page) == pageSet.end()) {
            if (pageFrames.size() == frameCount) {
                int removedPage = pageFrames.front();
                pageFrames.pop_front();
                pageSet.erase(removedPage);
            }
            pageFrames.push_back(page);
            pageSet.insert(page);
            pageFaults++;
            
            for (int frame : pageFrames) {
                cout << "       " << frame;
            }
            
            for (int j = pageFrames.size(); j < frameCount; j++) {
                cout << "                ";
            }

            cout << "| Page fault (F: " << pageFaults << ") |" << endl;
        } else {
            pageFrames.remove(page);
            pageFrames.push_back(page);
            pageHits++;
            
            for (int frame : pageFrames) {
                cout << "       " << frame;
            }
            
            for (int j = pageFrames.size(); j < frameCount; j++) {
                cout << "                ";
            }

            cout << "| Page hit (H: " << pageHits << ")   |" << endl;
        }
    }

    cout << "+-----+--------------+----------------+----------------+" << endl;
    cout << "| Total Page Faults: " << pageFaults;
    for (int j = to_string(pageFaults).length(); j < 8; j++) {
        cout << " ";
    }
    cout << " |" << endl;
    cout << "+-----------------------------+";
    for (int j = to_string(pageFaults).length(); j < 8; j++) {
        cout << " ";
    }
    cout << "+" << endl;
}

int main() {
    vector<int> pageRequests = {1, 3, 0, 3, 5, 6, 3};
    int frameCount = 3;

    simulateLRUPageReplacement(pageRequests, frameCount);

    return 0;
}
